File: DotNet/LruCache.Tests/ScalableLRUCacheTests.cs
namespace LRUCache.Tests;

public class ScalableLRUCacheTests
{
    [Fact(DisplayName = "Keys survive a resize that keeps total capacity the same")]
    public void Resize_SameCapacity_AllKeysSurvive()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 8, stripeCount: 2);
        for (int i = 0; i < 8; i++) cache.Put(i, i * 10);

        cache.Resize(newTotalCapacity: 8, newStripeCount: 4);

        for (int i = 0; i < 8; i++)
            Assert.True(cache.TryGet(i, out var v) && v == i * 10,
                $"Key {i} should survive resize with same capacity");
    }

    [Fact(DisplayName = "Shrink evicts coldest keys when capacity is reduced")]
    public void Resize_SmallerCapacity_HottestKeysSurvive()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 6, stripeCount: 2);
        cache.Put(1, 1);
        cache.Put(2, 2);
        cache.Put(3, 3);
        cache.Put(4, 4);
        cache.Put(5, 5);
        cache.Put(6, 6);

        // Access 4, 5, 6 to make them hot
        cache.TryGet(4, out _);
        cache.TryGet(5, out _);
        cache.TryGet(6, out _);

        cache.Resize(newTotalCapacity: 3, newStripeCount: 2);

        // Cold keys (1, 2, 3) should be evicted; hot keys (4, 5, 6) should survive
        Assert.False(cache.TryGet(1, out _), "Cold key 1 should be evicted");
        Assert.False(cache.TryGet(2, out _), "Cold key 2 should be evicted");
        Assert.False(cache.TryGet(3, out _), "Cold key 3 should be evicted");
        Assert.True(cache.TryGet(4, out _),  "Hot key 4 should survive");
        Assert.True(cache.TryGet(5, out _),  "Hot key 5 should survive");
        Assert.True(cache.TryGet(6, out _),  "Hot key 6 should survive");
    }

    [Fact(DisplayName = "Concurrent reads and writes do not corrupt state during normal operation")]
    public void Concurrent_ReadWrite_NoCorruption()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 16, stripeCount: 4);
        for (int i = 0; i < 16; i++) cache.Put(i, i);

        var errors = new System.Collections.Concurrent.ConcurrentBag<string>();

        var readers = Enumerable.Range(0, 8).Select(_ => Task.Run(() =>
        {
            for (int i = 0; i < 1000; i++)
            {
                int key = i % 16;
                if (cache.TryGet(key, out var v) && v != key)
                    errors.Add($"Key {key} returned wrong value {v}");
            }
        }));

        var writers = Enumerable.Range(0, 4).Select(_ => Task.Run(() =>
        {
            for (int i = 0; i < 500; i++)
                cache.Put(i % 16, i % 16);
        }));

        Task.WaitAll(readers.Concat(writers).ToArray());

        Assert.Empty(errors);
    }

    [Fact(DisplayName = "Put and Get work correctly after resize")]
    public void Resize_ThenPutGet_WorksCorrectly()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 4, stripeCount: 2);
        cache.Put(1, 10);
        cache.Put(2, 20);

        cache.Resize(newTotalCapacity: 8, newStripeCount: 4);

        cache.Put(3, 30);
        cache.Put(4, 40);

        Assert.True(cache.TryGet(1, out var v1) && v1 == 10);
        Assert.True(cache.TryGet(2, out var v2) && v2 == 20);
        Assert.True(cache.TryGet(3, out var v3) && v3 == 30);
        Assert.True(cache.TryGet(4, out var v4) && v4 == 40);
    }

    [Fact(DisplayName = "Count never exceeds total capacity after repeated resize")]
    public void RepeatedResize_CountNeverExceedsCapacity()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 8, stripeCount: 2);
        for (int i = 0; i < 8; i++) cache.Put(i, i);

        cache.Resize(newTotalCapacity: 4, newStripeCount: 2);
        for (int i = 0; i < 10; i++) cache.Put(i, i);
        Assert.True(cache.Count <= 4, $"After shrink: Count {cache.Count} > 4");

        cache.Resize(newTotalCapacity: 12, newStripeCount: 4);
        for (int i = 0; i < 20; i++) cache.Put(i, i);
        Assert.True(cache.Count <= 12, $"After grow: Count {cache.Count} > 12");
    }

    [Fact(DisplayName = "Resize while under concurrent load does not lose all keys")]
    public void Resize_UnderConcurrentLoad_CacheRemainsUsable()
    {
        var cache = new ScalableLRUCache<int, int>(totalCapacity: 8, stripeCount: 2);
        for (int i = 0; i < 8; i++) cache.Put(i, i);

        var resizeTask = Task.Run(() =>
            cache.Resize(newTotalCapacity: 8, newStripeCount: 4));

        var readTask = Task.Run(() =>
        {
            for (int i = 0; i < 1000; i++)
                cache.TryGet(i % 8, out _); // should not throw
        });

        Task.WaitAll(resizeTask, readTask);

        // Cache must still be functional after resize under load
        cache.Put(99, 99);
        Assert.True(cache.TryGet(99, out var v) && v == 99,
            "Cache must accept writes and return correct values after concurrent resize");
    }
}

File: Java/src/test/java/cache/ScalableLRUCacheTests.java
package cache;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Phase 5 — Scalable LRU cache: resize and concurrent correctness.
 *
 * These tests verify stripe-based partitioning, resize correctness under
 * quiescence, and safety under concurrent load.
 */
@Tag("Phase5")
class ScalableLRUCacheTests {

    @Test
    @DisplayName("same-capacity resize preserves all keys")
    void resize_SameCapacity_PreservesAllKeys() {
        var cache = new ScalableLRUCache<Integer, Integer>(100, 4);
        for (int i = 0; i < 80; i++) cache.put(i, i * 10);

        cache.resize(100, 8); // same total capacity, more stripes

        for (int i = 0; i < 80; i++)
            assertEquals(i * 10, cache.get(i),
                    "Key " + i + " missing after resize");
    }

    @Test
    @DisplayName("shrink evicts the coldest keys")
    void resize_Shrink_EjectsColdestKeys() {
        var cache = new ScalableLRUCache<Integer, Integer>(100, 4);
        for (int i = 0; i < 100; i++) cache.put(i, i);

        // Access keys 0-9 to make them hot
        for (int i = 0; i < 10; i++) cache.get(i);

        cache.resize(40, 4); // shrink — only 40 slots survive

        assertTrue(cache.size() <= 40, "size exceeds new capacity after shrink");

        // Hot keys should have a higher survival rate than cold ones
        long hotSurvived = 0;
        for (int i = 0; i < 10; i++) if (cache.get(i) != null) hotSurvived++;
        // At least some hot keys should survive
        assertTrue(hotSurvived > 0, "No hot keys survived the shrink");
    }

    @Test
    @DisplayName("concurrent reads and writes do not corrupt the cache")
    void concurrentReadsAndWrites_DoNotCorrupt() throws Exception {
        var cache = new ScalableLRUCache<Integer, Integer>(200, 4);
        for (int i = 0; i < 100; i++) cache.put(i, i);

        var error = new AtomicBoolean(false);
        runConcurrently(8, threadId -> {
            try {
                ThreadLocalRandom rng = ThreadLocalRandom.current();
                for (int i = 0; i < 3000; i++) {
                    int key = rng.nextInt(200);
                    if (rng.nextBoolean()) cache.put(key, key);
                    else cache.get(key);
                }
            } catch (Exception e) {
                error.set(true);
            }
        });

        assertFalse(error.get(), "Exception thrown during concurrent access");
        assertTrue(cache.size() <= 200, "size exceeded capacity");
    }

    @Test
    @DisplayName("resize under concurrent read load leaves cache usable")
    void resize_UnderConcurrentLoad_CacheRemainsUsable() throws Exception {
        var cache  = new ScalableLRUCache<Integer, Integer>(200, 4);
        for (int i = 0; i < 100; i++) cache.put(i, i);

        var error   = new AtomicBoolean(false);
        var latch   = new CountDownLatch(1);
        var executor = Executors.newFixedThreadPool(5);
        List<Future<?>> futures = new ArrayList<>();

        // 4 reader threads
        for (int t = 0; t < 4; t++) {
            futures.add(executor.submit(() -> {
                latch.await();
                try {
                    ThreadLocalRandom rng = ThreadLocalRandom.current();
                    for (int i = 0; i < 2000; i++) cache.get(rng.nextInt(100));
                } catch (Exception e) {
                    error.set(true);
                }
                return null;
            }));
        }

        // 1 resize thread
        futures.add(executor.submit(() -> {
            latch.await();
            try {
                cache.resize(200, 8);
            } catch (Exception e) {
                error.set(true);
            }
            return null;
        }));

        latch.countDown();
        executor.shutdown();
        assertTrue(executor.awaitTermination(30, TimeUnit.SECONDS));
        for (Future<?> f : futures) f.get();

        assertFalse(error.get(), "Exception thrown during resize under concurrent load");

        // Cache must still be functional after resize
        cache.put(999, 999);
        assertEquals(999, cache.get(999));
    }

    // ── Helpers ────────────────────────────────────────────────────────────

    @FunctionalInterface
    interface ThreadTask {
        void run(int threadId) throws Exception;
    }

    private void runConcurrently(int threadCount, ThreadTask task) throws Exception {
        var executor = Executors.newFixedThreadPool(threadCount);
        var latch    = new CountDownLatch(1);
        List<Future<?>> futures = new ArrayList<>();

        for (int t = 0; t < threadCount; t++) {
            final int id = t;
            futures.add(executor.submit(() -> {
                latch.await();
                task.run(id);
                return null;
            }));
        }

        latch.countDown();
        executor.shutdown();
        assertTrue(executor.awaitTermination(30, TimeUnit.SECONDS), "Threads did not finish in time");
        for (Future<?> f : futures) f.get();
    }
}
